confgenerator/confgenerator.go

```diff
@@ -97,6 +99,152 @@ func (m *Metrics) generateOtelPipelines() (map[string]otel.Pipeline, error) {
 	return out, nil
 }
 
+func (uc *UnifiedConfig) GenerateFluentBitConfigsModular(logsDir string, stateDir string, hostInfo *host.InfoStat) (string, string, error) {
+	userAgent, _ := getUserAgent("Google-Cloud-Ops-Agent-Logging", hostInfo)
+	components, err := uc.Logging.generateFluentbitComponents(userAgent, hostInfo)
+	if err != nil {
+		return "", "", err
+	}
+
+	c := fluentbit.ModularConfig{
+		Variables: map[string]string{
+			"buffers_dir": path.Join(stateDir, "buffers"),
+			"logs_dir":    logsDir,
+		},
+		Components: components,
+	}
+	return c.Generate()
+}
+
+type fbSource struct {
+	tag        string
+	components []fluentbit.Component
+}
+
+func (l *Logging) generateFluentbitComponents(userAgent string, hostInfo *host.InfoStat) ([]fluentbit.Component, error) {
+	var out []fluentbit.Component
+	out = append(out, fluentbit.Service{}.Component())
+
+	if l != nil && l.Service != nil {
+		var sources []fbSource
+		var logNames []string
+		for pID, p := range l.Service.Pipelines {
+			for _, rID := range p.ReceiverIDs {
+				receiver, ok := l.Receivers[rID]
+				if !ok {
+					return nil, fmt.Errorf("receiver %q not found", rID)
+				}
+				tag := fmt.Sprintf("%s.%s", pID, rID)
+				components := receiver.Components(tag)
+				for i, pID := range p.ProcessorIDs {
+					processor, ok := l.Processors[pID]
+					if !ok {
+						processor, ok = LegacyBuiltinProcessors[pID]
+					}
+					if !ok {
+						return nil, fmt.Errorf("processor %q not found", pID)
+					}
+					components = append(components, processor.Components(tag, i)...)
+				}
+				components = append(components, setLogNameComponents(tag, rID)...)
+				logNames = append(logNames, regexp.QuoteMeta(rID))
+				sources = append(sources, fbSource{tag, components})
+			}
+		}
+		sort.Slice(sources, func(i, j int) bool { return sources[i].tag < sources[j].tag })
+		sort.Strings(logNames)
+
+		for _, s := range sources {
+			out = append(out, s.components...)
+		}
+		if len(logNames) > 0 {
+			out = append(out, fluentbit.Stackdriver{
+				Match:     strings.Join(logNames, "|"),
+				Workers:   8,
+				UserAgent: userAgent,
+			}.Component())
+		}
+	}
+	// TODO: Use receivers instead of generating Tail objects directly.
+	if hostInfo.OS != "windows" {
+		out = append(out, fluentbit.Tail{
+			Tag:          "ops-agent-collectd",
```

Comment:
[Optional because it's not introduced by this PR] Should this and line 180 be updated to `ops-agent-open-telemetry-collector` instead now? In theory it's no-op now because otel writes to syslog instead of `metrics-module.log`.

Reply:
The OT collector does not write its logs to disk. We could technically delete this — the only reason I left it around is as a reminder to get the OT collector to write the logs here.

Reply:
Deleting it sounds good to me too.

Reply:
I've been preserving it at Igor's requests, but I'm just as happy to delete it.

Reply:
Removed.